Bluetooth: controller: Support user defined CPR intervals

thoh-ot opened the pull request:
Some proprietary protocols that reuse the Bluetooth LLCP logic requires that the CPR interval has a valid range that differs from the Bluetooth specification defined.

Vendor must implement the function (ull_vendor.h)
`u16_t ull_get_cpr_interval_min_limit(struct ll_conn *conn)`
that returns the valid limit for the given connection.

thoh-ot committed:
Bluetooth: controller: User defined CPR intervals

Some proprietary protocols that reuse the Bluetooth LLCP logic requires
that the CPR interval has a valid range that differs from the Bluetooth
specification defined.

Vendor must implement the function (ull_vendor.h)
uint16_t ull_conn_interval_min_get(struct ll_conn *conn)
that returns the valid limit for the given connection.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: User defined CPR intervals

Enable feature and add simple stub for compile test.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -51,6 +51,18 @@
+ */
+#if !defined(CONFIG_BT_CTLR_USER_CPR_INTERVAL)
+/* Bluetooth defined CPR Interval Minimum (7.5ms) */
+#define CPR_INTERVAL_MIN_LIMIT(x) (6)
```
```suggestion
#define CONN_INTERVAL_MIN(x) (6)
```

thoh-ot replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -51,6 +51,18 @@
+#define CPR_INTERVAL_MIN_LIMIT(x) (6)
+#else /* CONFIG_BT_CTLR_USER_CPR_INTERVAL */
+/* Proprietary user defined CPR Interval Minimum */
+extern uint16_t ull_get_cpr_interval_min_limit(struct ll_conn *conn);
+#define CPR_INTERVAL_MIN_LIMIT(x) (ull_get_cpr_interval_min_limit(x))
```
```suggestion
extern uint16_t ull_conn_interval_min_get(struct ll_conn *conn);
#define CONN_INTERVAL_MIN(x) (ull_conn_interval_min_get(x))
```

thoh-ot replied:
Done

thoh-ot commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -547,6 +547,13 @@ config BT_RX_USER_PDU_LEN
 	  header and payload. It does not account for HCI event headers as these
 	  PDUs are assumed to not go across HCI.
 
+config BT_CTLR_USER_CPR_INTERVAL
```
Well it was supposed to be a feature flag for user defined CPR intervals not only the minimum limit, but we have only implemented the minimum limit. So that we don't end up with a BT_CTLR_USER_CPR_INTERVAL_MAX also in the future.

wopu-ot replied:
I would prefer to have the name of the feature flag and its implementation in agreement. We can drop the `_MIN` suffix if we implement a maximum limit in the future.

thoh-ot replied:
@cvinayak Any comments regarding this change?

cvinayak replied:
I will go with @wopu-ot  and add the `_MIN` suffix, saves us from future rename if we want to set max value.

thoh-ot replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -51,6 +51,18 @@
+#else /* CONFIG_BT_CTLR_USER_CPR_INTERVAL */
+/* Proprietary user defined CPR Interval Minimum */
+extern uint16_t ull_conn_interval_min_get(struct ll_conn *conn);
+#define CONN_INTERVAL_MIN(x) (ull_conn_interval_min_get(x))
```
```suggestion
#define CONN_INTERVAL_MIN(x) MAX(ull_conn_interval_min_get(x), 6)
```

thoh-ot replied:
The purpose is to allow a proprietary protocol a connection interval that is outside the valid range so also less than 6. Maybe it wasn't clear but if you enabled the feature the ull_conn_interval_min_get function would have to look at the given connection and return a suitable minimum connection interval, so 6 for a proper Bluetooth LE connection and maybe something less for a proprietary protocol.

cvinayak replied:
Fine by me. Could you check if controller can gracefully handle the value 0 as the minimum interval?

thoh-ot replied:
Clamped it to a minimum of 1.

cvinayak reviewed (changes requested):
Shouldnt we restrict the minimum in the upstream code?

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -547,6 +547,13 @@ config BT_RX_USER_PDU_LEN
 	  PDUs are assumed to not go across HCI.
 
+config BT_CTLR_USER_CPR_INTERVAL
+	bool "Enable proprietary Connection Parameter Request intervals"
```
```suggestion
	bool "Enable proprietary Connection Parameter Request minimum interval"
```

thoh-ot replied:
Done

thoh-ot commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -547,6 +547,13 @@ config BT_RX_USER_PDU_LEN
+	depends on BT_CTLR_USER_EXT
+	help
+	  When enabled, controller will accept proprietary values for the
+	  Connection Parameter Request intervals.
```
Done

thoh-ot commented:
There is an unrelated build error in CI due to ull_sched.c missing to include a dependency for lll.h to able to work.

thoh-ot committed:
Bluetooth: controller: Fix header inclusion

Use correct headers to make the inclusion of lll.h work when
CONFIG_BT_CTLR_USER_EXT is enabled.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>
